File: Ahorn/entities/dreamDashBerry.jl
module CommunalHelperDreamStrawberry

using ..Ahorn, Maple

@mapdef Entity "CommunalHelper/DreamStrawberry" DreamDashBerry(
	x::Integer,
	y::Integer,
	order::Integer=-1,
	checkpointID::Integer=-1,
)

const placements = Ahorn.PlacementDict(
	"Dream Strawberry (Communal Helper)" => Ahorn.EntityPlacement(
		DreamDashBerry
	),
)

const sprite = "collectables/CommunalHelper/dreamberry/wings01"
const seedSprite = "collectables/CommunalHelper/dreamberry/seed02"

Ahorn.nodeLimits(entity::DreamDashBerry) = 0, -1

function Ahorn.selection(entity::DreamDashBerry)
    x, y = Ahorn.position(entity)

    res = Ahorn.Rectangle[Ahorn.getSpriteRectangle(sprite, x, y)]
    
    nodes = get(entity.data, "nodes", ())
    for node in nodes
        nx, ny = node
        push!(res, Ahorn.getSpriteRectangle(seedSprite, nx, ny))
    end

    return res
end

function Ahorn.renderSelectedAbs(ctx::Ahorn.Cairo.CairoContext, entity::DreamDashBerry)
    x, y = Ahorn.position(entity)

    for node in get(entity.data, "nodes", ())
        nx, ny = node
        Ahorn.drawLines(ctx, Tuple{Number, Number}[(x, y), (nx, ny)], Ahorn.colors.selection_selected_fc)
    end
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::DreamDashBerry, room::Maple.Room)
	x, y = Ahorn.position(entity)
	Ahorn.drawSprite(ctx, sprite, x, y)

    nodes = get(entity.data, "nodes", ())
    for node in nodes
        nx, ny = node
        Ahorn.drawSprite(ctx, seedSprite, nx, ny)
    end
end

end

File: Loenn/entities/dream_dash_berry.lua
local utils = require("utils")

local dreamDashBerry = {}

dreamDashBerry.name = "CommunalHelper/DreamStrawberry"
dreamDashBerry.depth = -100
dreamDashBerry.nodeLineRenderType = "fan"
dreamDashBerry.nodeLimits = {0, -1}

dreamDashBerry.fieldInformation = {
    order = {
        fieldType = "integer"
    },
    checkpointID = {
        fieldType = "integer"
    }
}

dreamDashBerry.placements = {
    name = "dream_dash_berry",
    data = {
        order = -1,
        checkpointID = -1
    }
}

dreamDashBerry.texture = "collectables/CommunalHelper/dreamberry/wings01"
dreamDashBerry.nodeTexture = "collectables/CommunalHelper/dreamberry/seed02"

function dreamDashBerry.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local nodes = entity.nodes or {{x = 0, y = 0}}

    local rects = {}
    for _, node in ipairs(nodes) do
        local nx, ny = node.x or 0, node.y or 0
        table.insert(rects, utils.rectangle(nx - 4, ny - 4, 7, 9))
    end

    return utils.rectangle(x - 6, y - 7, 12, 13), rects
end

return dreamDashBerry
